samples/mqtt/x509_credentials_provider_connect/main.cpp

```diff
@@ -0,0 +1,247 @@
+    String x509CertificatePath = cmdUtils.GetCommandRequired(
+        "x509_cert", "X509 credentials sourcing requires an IoT certificate to be specified.");
+    String x509KeyPath = cmdUtils.GetCommandRequired(
+        "x509_key", "X509 credentials sourcing requires an IoT thing private key to be specified.");
+    String x509RootCAFile = cmdUtils.GetCommandOrDefault("x509_ca_file", "");
+
+    /********************** Now Setup an Mqtt Client ******************/
+    Aws::Crt::Io::TlsContext x509TlsCtx;
+    Aws::Iot::MqttClientConnectionConfigBuilder builder;
+    std::shared_ptr<Aws::Crt::Auth::ICredentialsProvider> provider = nullptr;
+
+    Aws::Crt::Io::TlsContextOptions tlsCtxOptions =
+        Aws::Crt::Io::TlsContextOptions::InitClientWithMtls(x509CertificatePath.c_str(), x509KeyPath.c_str());
+    if (!tlsCtxOptions)
+    {
+        fprintf(
+            stderr,
+            "Unable to initialize tls context options, error: %s!\n",
+            ErrorDebugString(tlsCtxOptions.LastError()));
+        return -1;
+    }
+
+    if (!x509RootCAFile.empty())
+    {
+        tlsCtxOptions.OverrideDefaultTrustStore(nullptr, x509RootCAFile.c_str());
+    }
+
+    x509TlsCtx = Aws::Crt::Io::TlsContext(tlsCtxOptions, Aws::Crt::Io::TlsMode::CLIENT);
+    if (!x509TlsCtx)
+    {
+        fprintf(
+            stderr,
+            "Unable to create tls context, error: %s!\n",
+            ErrorDebugString(x509TlsCtx.GetInitializationError()));
+        return -1;
+    }
+
+    Aws::Crt::Auth::CredentialsProviderX509Config x509Config;
+
+    x509Config.TlsOptions = x509TlsCtx.NewConnectionOptions();
+    if (!x509Config.TlsOptions)
+    {
+        fprintf(
+            stderr,
+            "Unable to create tls options from tls context, error: %s!\n",
+            ErrorDebugString(x509Config.TlsOptions.LastError()));
+        return -1;
+    }
+
+    x509Config.Endpoint = x509Endpoint;
+    x509Config.RoleAlias = x509RoleAlias;
+    x509Config.ThingName = x509ThingName;
+
+    Aws::Crt::Http::HttpClientConnectionProxyOptions proxyOptions;
+    if (!proxyHost.empty())
+    {
+        proxyOptions.HostName = proxyHost;
+        proxyOptions.Port = proxyPort;
+        proxyOptions.AuthType = Aws::Crt::Http::AwsHttpProxyAuthenticationType::None;
+        x509Config.ProxyOptions = proxyOptions;
+    }
+
+    provider = Aws::Crt::Auth::CredentialsProvider::CreateCredentialsProviderX509(x509Config);
+
+    if (!provider)
+    {
+        fprintf(stderr, "Failure to create credentials provider!\n");
+        return -1;
+    }
+
+    Aws::Iot::WebsocketConfig config(signingRegion, provider);
+    builder = Aws::Iot::MqttClientConnectionConfigBuilder(config);
+
+    if (!proxyHost.empty())
+    {
+        builder.WithHttpProxyOptions(proxyOptions);
+    }
+
+    if (!caFile.empty())
+    {
+        builder.WithCertificateAuthority(caFile.c_str());
+    }
+
+    builder.WithEndpoint(endpoint);
+
+    auto clientConfig = builder.Build();
```

Comment:
For the connect samples, isn't everything the same after the client config construction?  Can we refactor that to not be repeated?

Along the same lines, why don't these samples use the utils connection builder helper functions?